applications/nrf_desktop/doc/hid_provider_keyboard.rst

```diff
@@ -0,0 +1,136 @@
+For backwards compatibility, the default value for this option can be set using the deprecated :ref:`CONFIG_DESKTOP_HID_EVENT_QUEUE_SIZE <config_desktop_app_options>` Kconfig option.
+If there is no space in the queue to enqueue a new key state change, the oldest element is released.
+
+Report expiration
+-----------------
+
+With the :ref:`CONFIG_DESKTOP_HID_REPORT_PROVIDER_KEYBOARD_KEYPRESS_EXPIRATION <config_desktop_app_options>` Kconfig option, you can set the amount of time after which a queued key will be considered expired.
+For backwards compatibility, the default value for this option can be set using the deprecated :ref:`CONFIG_DESKTOP_HID_REPORT_EXPIRATION <config_desktop_app_options>` Kconfig option.
+The higher the value, the longer the period from which the nRF Desktop application will recall pressed keys when the connection with HID host is established.
+
+Handling keys state
+===================
+
+The module selects the :ref:`CONFIG_DESKTOP_KEYS_STATE <config_desktop_app_options>` Kconfig option to enable the :ref:`nrf_desktop_keys_state`.
+The utility is used to track the state of active keys after the connection with the HID host is established.
+
+Implementation details
+**********************
+
+On initialization, the module announces its presence by sending :c:struct:`hid_report_provider_event` event with appropriate report ID and implementation of callbacks from HID report provider API (:c:struct:`hid_report_provider_api`).
+Separate :c:struct:`hid_report_provider_event` events are sent for each report ID.
+The module supports keyboard reports from Report protocol and Boot protocol which can be enabled using :ref:`CONFIG_DESKTOP_HID_BOOT_INTERFACE_KEYBOARD <config_desktop_app_options>`.
+The :ref:`nrf_desktop_hid_state` receives the events and fills them with its own implementation of callbacks from the HID state API (:c:struct:`hid_state_api`).
+After that process the modules can communicate by calling each other callbacks.
+The module also subscribes to :c:struct:`button_event` events to get information about the button presses.
+The module sends :c:struct:`hid_report_event` events to appropriate subscriber when it is requested by the :ref:`nrf_desktop_hid_state`.
+
+Linking input data with the right HID report
+============================================
+
+Out of all available input data types, the module collects button events.
+The button events are stored in the :c:struct:`report_data` structure.
+The ``button_event`` is the source of this type of data.
+
+To indicate a change to this input data, the module overwrites the value that is already stored.
+
+Since keys on the board can be associated to a HID usage ID, and thus be part of different HID reports, the first step is to identify if the key belongs to a HID report that is provided by this module.
```

Comment:
```suggestion
Since keys on the board can be associated with a HID usage ID and thus be part of different HID reports, the first step is to identify if the key belongs to a HID report that is provided by this module.
```